File: data-structures/singly-linked-list/MyLinkedList.java
// Java implementation of Singly linked list 

class Node {
    int data;
    Node next;

    public Node(int data) {
        this.data = data;
        this.next = null;
    }
}

public class MyLinkedList {
    private Node head;

    public MyLinkedList() {
        head = null;
    }

    // Insert 'value' at the front of the list
    public void insertAtFront(int value) {
        Node node = new Node(value);
        if(isEmpty()) {
            head = node;
        } else {
            node.next = head;
            head = node;
        }
    }

    // insert value at the back of the linked list
    public void insertAtBack(int value) {
        Node node = new Node(value);
        if (isEmpty()) {
            head = node;
        } else {
            Node curr = head;
            while(null != curr.next) {
                curr = curr.next;
            }

            curr.next = node;
        }
    }

    // inserts value after key
    public void insertAfter(int key, int value) {
        Node node = new Node(value);

        // find the position of key
        Node curr = head;
        while(null != curr && curr.data != key) {
            curr = curr.next;
        }

        if (null == curr) {
            System.out.println("Key not found");
            return;
        }

        if(null == curr.next) {
            curr.next = node;
        } else {
            Node next = curr.next;
            curr.next = node;
            node.next = next;
        }
    }

    // returns the data at first node 
    public int topFront() {
        if (isEmpty()) {
            System.out.println("List is empty");
            return -999999;
        }

        return head.data;
    }  

    // returns the data at last node 
    public int topBack() { 
        if(isEmpty()) {
            System.out.println("List is empty");
            return -999999;
        }

        Node curr = head;
        while(curr.next != null) {
            curr = curr.next;
        }

        return curr.data;
    }

    // removes the item at front of the linked list and return 
    public int popFront() {
        if (isEmpty()) {
            System.out.println("List is empty");
            return -999999;
        }

        Node next = head.next;
        int item = head.data;
        head = next;
        return item;
    }

    // remove the item at the end of the list and return 
    public int popBack() {
        if (isEmpty()) {
            System.out.println("List is empty");
            return -999999;
        }

        // get to the last node
        Node curr = head;
        Node prev = null;
        while(curr.next != null) {
            prev = curr;
            curr = curr.next;
        }

        int item = curr.data;

        if (prev == null) {
            // the list has only one item
            head = null;
        } else {
            prev.next = null;
            curr = null;
        }

        return item;
    }

    // removes an item with value 'key'
    public void remove(int key) {
        if (isEmpty()) {
            System.out.println("List is empty");
            return;
        }

        // find the position of the key
        Node curr = head;
        Node prev = null;

        while(curr != null && curr.data != key) {
            prev = curr;
            curr = curr.next;
        }

        if(curr == null) {
            System.out.print("key not found");
            return;
        }

        // if curr is head, delete the head
        if (prev == null) {
            head = head.next;
            curr = null;
        } else if (curr.next == null) { // if curr is last item
            prev.next = null;
            curr = null;
        } else { //anywhere between first and last node
            Node next = curr.next;
            prev.next = next;
            curr = null;
        }   


    }

    // check if the key is in the list
    public boolean find(int key) {
        if (isEmpty()) {
            System.out.println("List is empty");
            return false;
        }

        Node curr = head;
        while(curr != null && curr.data != key) {
            curr = curr.next;
        }

        if (curr == null) {
            return false;
        }

        return true;
    }

    // check if the list is empty
    public boolean isEmpty() {
        return head == null;
    }

    // print all the items
    public void print() {
        if (isEmpty()) {
            System.out.println("Nothing to display");
        } else {
            Node curr = head;
            while (curr != null) {
                System.out.print(curr.data + " ");
                curr = curr.next;
            }
            System.out.println();
        }
    }

    public static void main(String [] args) {
        MyLinkedList list = new MyLinkedList();
    }
}

File: data-structures/singly-linked-list/my_linked_list.py
# Python implementation of Singly linked list 
import sys

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class MyLinkedList:
    def __init__(self):
        self.head = None 

    # Insert 'value' at the front of the list
    def insert_at_front(self, value):
        node = Node(value)
        if self.is_empty():
            self.head = node
        else:
            node.next = self.head
            self.head = node
     
    #  insert value at the back of the linked list
    def insert_at_back(self, value):
        node = Node(value)
        if self.is_empty():
            self.head = node
        else:
            curr = self.head
            while curr.next != None:
                curr = curr.next
            curr.next = node

    # inserts value after key
    def insert_after(self, key, value):
        node = Node(value)

        # find the position of key
        curr = self.head
        while curr != None and curr.data != key:
            curr = curr.next

        if curr == None:
            print 'Key not found'
            return

        if curr.next == None:
            curr.next = node
        else:
            next_node = curr.next
            curr.next = node
            node.next = next_node

    # returns the data at first node 
    def top_front(self):
        if self.is_empty():
            print 'List is empty'
            return

        return self.head.data

    # returns the data at last node 
    def top_back(self): 
        if self.is_empty():
            print 'List is empty'
            return

        curr = self.head
        while curr.next != None:
            curr = curr.next

        return curr.data

    # removes the item at front of the linked list and return 
    def pop_front(self):
        if self.is_empty():
            print 'List is empty'
            return

        next_node = self.head.next
        item = self.head.data
        self.head = next_node
        return item

    # remove the item at the end of the list and return 
    def pop_back(self):
        if self.is_empty():
            print 'List is empty'
            return

        # get to the last node
        curr = self.head
        prev = None

        while curr.next != None:
            prev = curr
            curr = curr.next

        item = curr.data

        if prev == None:
            # the list has only one item
            self.head = None
        else:
            prev.next = None
            curr = None

        return item

    # removes an item with value 'key'
    def remove(self, key):
        if self.is_empty():
            print 'List is empty'
            return

        # find the position of the key
        curr = self.head
        prev = None

        while curr != None and curr.data != key:
            prev = curr
            curr = curr.next

        if curr == None:
            print 'key not found'
            return

        # if curr is head, delete the head
        if prev == None:
            self.head = self.head.next
            curr = None
        elif curr.next == None: # if curr is last item
            prev.next = None
            curr = None
        else: #anywhere between first and last node
            next_node = curr.next
            prev.next = next_node
            curr = None

    # check if the key is in the list
    def find(self, key):
        if self.is_empty():
            print 'List is empty'
            return False

        curr = self.head
        while curr != None and curr.data != key:
            curr = curr.next

        if curr == None:
            return False

        return True

    # check if the list is empty
    def is_empty(self):
        return self.head == None

    # print all the items
    def printlist(self):
        if self.is_empty():
            print "Nothing to display"
        else:
            curr = self.head
            while (curr != None):
                sys.stdout.write(str(curr.data) + ' ')
                curr = curr.next

            print ''
       
if __name__ == '__main__':
    list = MyLinkedList()
    # write code